packages/base/src/widget.ts

```diff
@@ -219,64 +218,65 @@ export class WidgetModel extends Backbone.Model {
   /**
    * Handle incoming comm msg.
    */
-  _handle_comm_msg(msg: KernelMessage.ICommMsgMsg): Promise<void> {
+  async _handle_comm_msg(msg: KernelMessage.ICommMsgMsg): Promise<void> {
     const data = msg.content.data as any;
     const method = data.method;
     switch (method) {
       case 'update':
       case 'echo_update':
-        this.state_change = this.state_change
-          .then(() => {
-            const state: Dict<BufferJSON> = data.state;
-            const buffer_paths = data.buffer_paths ?? [];
-            const buffers = msg.buffers?.slice(0, buffer_paths.length) ?? [];
-            utils.put_buffers(state, buffer_paths, buffers);
-
-            if (msg.parent_header && method === 'echo_update') {
-              const msgId = (msg.parent_header as any).msg_id;
-              // we may have echos coming from other clients, we only care about
-              // dropping echos for which we expected a reply
-              const expectedEcho = Object.keys(state).filter((attrName) =>
-                this._expectedEchoMsgIds.has(attrName)
-              );
-              expectedEcho.forEach((attrName: string) => {
-                // Skip echo messages until we get the reply we are expecting.
-                const isOldMessage =
-                  this._expectedEchoMsgIds.get(attrName) !== msgId;
-                if (isOldMessage) {
-                  // Ignore an echo update that comes before our echo.
+        try {
+          await this.state_change;
```

Comment:
This one is a little tricky. We always want to process messages in order, but the message processing is async. So we have to queue any new message behind the other messages that have already come in. We do this creating a promise for every message we'd like to process. The `this.state_change` attribute points to the last message we've queued. So we chain our new message processing off of `this.state_change` (that's the `then`), and then we update `this.state_change` to point to the new promise we created. It's sort of like a linked list (i.e., a promise chain) where `this.state_change` always points to the last element of the chain, so we know what we need to chain the next promise off of.

Because we need to update `this.state_change`, I don't think we can convert this function to async/await. However, we might be able to refactor the bulk of this function into another async function, then call that to get the promise we need to update `this.state_change` with.

Reply:
That makes things clearer, I see. Seems like this change to [create view](https://github.com/jupyter-widgets/ipywidgets/pull/3584/commits/a843b1d5a29a0eb5e8476f5b852fbda4feb2bf00#diff-4fd512840569881075db4057ecca74fba82b67d649c904876c71d24d8764a8d2L154) in manager base should also remain unconverted to an async function?

Reply:
Updated with some of the logic refactored into an async helper method, and reverted the async change to create_view since this was another place where promises are chained to queue state changes. Tests passed locally.